### ISSUE TITLE
Non-ddev.site hostnames can't work on WSL2, and neither can ddev.site if DNS rebinding disallowed

### ISSUE THREAD

**Issue by rfay** [drud/ddev#2533]
**Describe the bug**

When the hostname required for a project is not *.ddev.site, or when DNS rebinding is not allowed, WSL2 browser stuff can't work.

So:
1. If DNS rebinding is disallowed, ddev will add the hostname to /etc/hosts. But unfortunately it's the /etc/hosts in the WSL2 system, which is *not* the hosts file that the browser will use (which is C:\Windows\System32\drivers\etc\hosts). So the browser can't recognize it.
2. If project_tld is set to something other than ddev.site, same problem
3. If additional_fqdns is set, that has the same problem.

I'm not sure there's a good solution to this, because WSL2 code can't change things on the Windows side...

Windows users can work around this issue by manually adding the needed hostnames to C:\Windows\System32\drivers\etc\hosts or by using the *windows* ddev to issue the command `ddev hostname <neededfqdn> 127.0.0.1`

**Comment by steffenf14** [comment]
It appears this issues just randomly appeared for me and colleges out of nowhere :| . Is there any way to allow this DNS rebinding you are talking about in 1. or won't this help?

**Comment by rfay** [comment]
@steffenf14 if DNS rebinding isn't allowed on your DNS server, you can:

1. Change your computer's DNS server to one that does allow DNS rebinding (Google's 8.8.8.8, 8.8.4.4 or Cloudfront's 1.1.1.1 for example) (Or change the DNS server handed out by your DHCP server)
2. Allow DNS rebinding (see https://github.com/drud/ddev/issues/2533)
3. Add hostnames on the Windows side either by editing the hosts file (above) or by using `ddev hostname` on the Windows side. Unfortunately, a hostname has to be added for each project/hostname.

**Comment by steffenf14** [comment]
@rfay I'm on a private home network with a pihole but that has been the case for a year. my College is also at his home but does use a default setup which also used to work like 3 days ago.

I just checked my /etc/hosts and the windows equivalent Windows/System32/driver/etc/hosts
DDEV appears to write my tests to only the WSL hosts and the windows file stays unaffected. This used to work like a week ago. Any ideas what (update) caused this problem?

**Comment by steffenf14** [comment]
this gets written to my WSL hosts file by ddev
127.0.0.1 test2.ddev.site test3.ddev.site test4 test5
the windows hosts file receives no such entries :/

**Comment by rfay** [comment]
That's exactly the problem in this issue - If DNS doesn't return the 127.0.0.1 value for *.ddev.site, ddev adds the hostname to the WSL2 hosts file, but if you're using a browser on Windows, then it doesn't know about the WSL2 hosts file.

**Comment by rfay** [comment]
Another option is to use a browser inside WSL2, which should work fine.  This is not hard to do, you just have to set up an X server like X411. Details on doing this for PHPStorm in https://ddev.com/ddev-local/ddev-local-and-phpstorm-debugging-with-wsl2/

**Comment by steffenf14** [comment]
I'm way more bothered by the fact that this used to work a week ago and I would like to find a solution to make this work again and not work around this bug.

**Comment by rfay** [comment]
I missed your comment in https://github.com/drud/ddev/issues/2533#issuecomment-710054884 that you're on a private network. 

Do you have a Fritzbox that allows turning off rebinding, see https://github.com/drud/ddev/issues/2409#issuecomment-686718237 ? Also, if you're running DNSMasq. 

A simple option is just to change your router (or your computer) to use a more generous DNS server like 8.8.8.8

**Comment by steffenf14** [comment]
Incredible good guess. Yes it's a FritzBox and I just recently received an update. 
Complete disableing of the DNS rebind protection is not possible but an exception rule fixed the issue for me. I just never realized that *.ddev.site is loopback address. Seems like *.ddev.site used to be tolerated by the my router but not anymore after the update.
Since I do not want to hardcode each and every domain into my fritzbox or hosts file i will switch to my piHole local DNS server directly.

But are there any plans for adding domains to windows hosts file from WSL ddev to counter this issue?

And thank you for your patience :D

**Comment by rfay** [comment]
I don't know of any way to have ddev inside WSL2 change the hosts file on Windows, but I'll keep thinking about that. The Windows OS is a completely different computer.  It's an interesting problem. Mostly people use forgiving DNS servers so this is not an issue so far, but I certainly consider it a very real problem.

**Comment by rfay** [comment]
At the time of https://github.com/drud/ddev/issues/2409#issuecomment-686718237 it was possible to remove the DNS rebinding from the FritzBox (see that comment). If that is no longer true, could you comment there?

I hope you'll also write up there your strategy with pi-hole... super interesting. It's essentially just short-circuiting DNS, and could work fine in this particular case.

**Comment by steffenf14** [comment]
just as a test, this works just fine:
`echo "127.0.0.1 test.ddev.site\r\n" >> $(wslpath "C:\Windows\System32\drivers\etc\hosts")`

**Comment by rfay** [comment]
Interesting, thanks. That has some potential, although ddev has to be add and remove hosts, and has always used a library to do that.

But how in the world can WSL2 edit an administrator-protected file like the hosts file?

**Comment by steffenf14** [comment]
@rfay only via elevated(administrator) shell sadly. I forgot to mention that sorry 🤐 my bad.

**Comment by rfay** [comment]
So you can use `sudo` or root user inside the container and fiddle with the Windows hosts file? Seems evil.

**Comment by steffenf14** [comment]
Agreed, that seems evil but I was just testing what is possible. 
Maybe something like a "WSL/Windows Companion Package for DDEV" that should be installed on Windows hosts (via choco/winget) that takes care of hosts entries. This could just be an optional package for DDEV-WSL users that want/need fqdns or have DNS rebinding issues.

**Comment by rfay** [comment]
That's an interesting strategy also.

Just to be clear: Were you able to solve the DNS problem so you don't have to manage hosts?

**Comment by steffenf14** [comment]
Yes, I was able to solve it for myself by switching to my PiHole DNS directly. 
My colleges will have to go the hard way of writing each entry to their windows hosts file for each project 😟 
Which will vary from 10 to 100+ local DNS Recods since our default domains are in this pattern `[projectName].[ourCompanyName]` . Sometimes with language prefix or subdomains.
We are currently transitioning from selfmade docker-compose files for each project to DDEV.

**Comment by rfay** [comment]
Why did this problem also occur for your colleagues? Are you using *.ddev.site? (Trying to understand the big picture)

**Comment by steffenf14** [comment]
I'm not very good at explaining but I'll try my best.
We are all in HomeOffice. So we have different Networks, so I cannot really speak for them, some may work fine some may not.
We thought our default domains (for example "customer.company" or "en.customer.company") worked just fine because with our old docker-compose setups we already had them listed in our Windows hosts file 🤐 .

I have been using DDEV for my personal projects which worked fine but stopped working after my router update, fixed that by switching DNS. For my personal projects, I used *.ddev.site domains or just plain "localhost" as FQDNS which obviously never conflicted with this issue.
At work, we just started switching some projects, but only recently with new projects, we encountered this issue because those domains were not listed in our windows hosts file. So we got really confused why some worked and some didn't and also some just worked only for others. We just didn't get the whole picture.
Thank you for your support!

**Comment by rfay** [comment]
If you're using ddev.site, I would expect your colleagues to do fine with WSL2. So far the DNS rebinding has *only* been reported by people with locked-down fritzbox. If you're *not* using ddev.site, then you can fix it with any number of DNS workarounds, but it's not as easy. I recommend using ddev.site.

You did understand that just putting "ddev.site' into your fritzbox config would fix it there right? But I guess you're using other FQDNs. companyname.ddev.site is far easier.

**Comment by wolcen** [comment]
Just to clarify: "So far the DNS rebinding has only been reported by people with locked-down fritzbox."
That may be the popular issue, but this certainly can happen with any DNS server that disables rebinding (a security feature for a very old problem).

In my case, a DNSMASQ server was doing this by default. In addition to fritzbox, this will also be the case on dd-wrt and openwrt, and likely others. Pi-hole should be fine since it won't work with dns rebind protection enabled.

**Comment by rfay** [comment]
@gilbertsoft suggests this script could be used from within WSL2 to run ddev on the WIndows side, and manage hosts there.

```
#!/usr/bin/env sh
#
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.
if [ "$DDEV_WSL_DEBUG_INFO" = true ]; then
	set -x
fi

NAME="ddev"
DDEV_PATH="$(dirname "$(realpath "$0")")"
DDEV="$DDEV_PATH/$NAME.exe"
SUDO="$DDEV_PATH/sudo.exe"

"$SUDO" "$DDEV" "$@"
exit $?

IN_WSL=false
if [ -n "$WSL_DISTRO_NAME" ]; then
	# $WSL_DISTRO_NAME is available since WSL builds 18362, also for WSL2
	IN_WSL=true
else
	WSL_BUILD=$(uname -r | sed -E 's/^[0-9.]+-([0-9]+)-Microsoft.*|.*/\1/')
	if [ -n "$WSL_BUILD" ]; then
		if [ "$WSL_BUILD" -ge 17063 ]; then
			# WSLPATH is available since WSL build 17046
			# WSLENV is available since WSL build 17063
			IN_WSL=true
		else
			# If running under older WSL, don't pass cli.js to Electron as
			# environment vars cannot be transferred from WSL to Windows
			# See: https://github.com/microsoft/BashOnWindows/issues/1363
			#      https://github.com/microsoft/BashOnWindows/issues/1494
			"$ELECTRON" "$@"
			exit $?
		fi
	fi
fi
if [ $IN_WSL = true ]; then

	export WSLENV="ELECTRON_RUN_AS_NODE/w:$WSLENV"
	CLI=$(wslpath -m "$VSCODE_PATH/resources/app/out/cli.js")

	# use the Remote WSL extension if installed
	WSL_EXT_ID="ms-vscode-remote.remote-wsl"

	ELECTRON_RUN_AS_NODE=1 "$ELECTRON" "$CLI" --locate-extension $WSL_EXT_ID >/tmp/remote-wsl-loc.txt 2>/dev/null
	WSL_EXT_WLOC=$(cat /tmp/remote-wsl-loc.txt)

	if [ -n "$WSL_EXT_WLOC" ]; then
		# replace \r\n with \n in WSL_EXT_WLOC
		WSL_CODE=$(wslpath -u "${WSL_EXT_WLOC%%[[:cntrl:]]}")/scripts/wslCode.sh
		"$WSL_CODE" "$COMMIT" "$QUALITY" "$ELECTRON" "$APP_NAME" "$DATAFOLDER" "$@"
		exit $?
	fi

elif [ -x "$(command -v cygpath)" ]; then
	CLI=$(cygpath -m "$VSCODE_PATH/resources/app/out/cli.js")
else
	CLI="$VSCODE_PATH/resources/app/out/cli.js"
fi
ELECTRON_RUN_AS_NODE=1 "$ELECTRON" "$CLI" "$@"
exit $?
```

**Comment by rfay** [comment]
Would appreciate review of #2605 by anyone in this issue, thanks!

**Comment by rfay** [comment]
Added some docs to help people sort this out. The docs will show up in v1.16; the fix (which @gilbertsoft is working hard on) may not show up until v1.17 or a point release of v1.16.

**Comment by rfay** [comment]
This needs to be addressed, and we may be able to use gsudo to do this, see https://github.com/gerardog/gsudo#usage-from-wsl-windows-subsystem-for-linux

**Comment by rfay** [comment]
This was resolved in DDEV v1.21.4, which now uses Windows-side hosts file management in this situation.
